docs/2022/puzzles/day10.md

```diff
@@ -6,6 +6,149 @@ import Solver from "../../../../../website/src/components/Solver.js"
+
+```scala
+  @main def part1(): Unit = println(s"The solution is ${registerStrengthsIterator.sum}")
+```
+
+### Part 2
+
+In the second part, we are asked to draw a CRT output. As stated in the puzzle description, the register is interpreted as the position of a the sprite `###`. The CRT iterates throught each line and, if the sprites touches the touches the current position, draws a `#`. Otherwise the CRT draws a `.`. The register's cycles are stepped in synced with the CRT.
+
+First, the CRT's position is just the cycle's index modulo the CRT's width (40 in our puzzle). Then, the CRT draw the sprite if and only if the register's value is the CRT's position, one more or one less. In other words, if `(reg_value - (cycle_id % 40)).abs <= 1`. Using the `zipWithIndex` method to obtain the cycles' indexes we end up with the following code:
+
+```scala
+  def CRTCharIterator: Iterator[Char] =
+    for ((reg, cycle) <- registerValuesIterator.zipWithIndex) yield {
+      if ((reg - (cycle % CRT_WIDTH)).abs <= 1) '#' else '.'
+    }
+```
+
+Now, we just need to concatenate the chars and add new lines at the right places. This is done using the `mkString` methods:
+
+```scala
+  def CRTImage: String = CRTCharIterator.grouped(CRT_WIDTH).map(_.mkString).mkString("\n")
+
+  @main def part2(): Unit = println(s"The CRT output is:\n$CRTImage")```
+
+## Final Code
+```scala
+package day10
+
+import locations.Directory.currentDir
+import inputs.Input.loadFileSync
+
+import Direction.*
+
+def INPUT(): String = loadFileSync(s"$currentDir/../input/day10")
+
+enum Command {
+  case NOOP
+  case ADDX(X: Int)
+}
+
+export Command.*
+
+def commandsIterator: Iterator[Command] = for (line <- INPUT.linesIterator) yield line.strip match {
+  case "noop" => NOOP
+  case s"addx $x" if x.toIntOption.isDefined => ADDX(x.toInt)
+  case _ => throw IllegalArgumentException(s"Invalid command '$line''")
+}
+
+val REGISTER_START_VALUE = 1
+
+def registerValuesIterator: Iterator[Int] = {
+  commandsIterator.scanLeft(REGISTER_START_VALUE :: Nil) {
+    case (_ :+ value, NOOP) => value :: Nil
+    case (_ :+ value, ADDX(x)) => value :: value + x :: Nil
+  }
+}.flatten
+
+def registerStrengthsIterator: Iterator[Int] = {
+  val it = for ((reg, i) <- registerValuesIterator.zipWithIndex) yield (i + 1) * reg
+  it.drop(19).grouped(40).map(_.head)
+}
+
+@main def part1(): Unit = println(s"The solution is ${registerStrengthsIterator.sum}")
+
+val CRT_WIDTH: Int = 40
+
+def CRTCharIterator: Iterator[Char] =
+  for ((reg, cycle) <- registerValuesIterator.zipWithIndex) yield {
+    if ((reg - (cycle % CRT_WIDTH)).abs <= 1) '#' else '.'
+  }
+
+def CRTImage: String = CRTCharIterator.grouped(CRT_WIDTH).map(_.mkString).mkString("\n")
+
+@main def part2(): Unit = println(s"The CRT output is:\n$CRTImage")
+```
+
+### Run it in the browser
+
+#### Part 1
+
+<Solver puzzle="day09-part1" year="2022"/>
+
+#### Part 2
+
+<Solver puzzle="day09-part2" year="2022"/>
```

Comment:
```suggestion
<Solver puzzle="day10-part2" year="2022"/>
```

Reply:
the puzzle key is significant because it is used to find the actual code to run the solution